src/daft-plan/src/treenode.rs

```diff
@@ -0,0 +1,85 @@
+use std::sync::Arc;
+
+use crate::{physical_plan::PhysicalPlan, LogicalPlan};
+use common_error::DaftResult;
+use common_treenode::{TreeNode, VisitRecursion};
+
+impl TreeNode for LogicalPlan {
+    fn apply_children<F>(&self, op: &mut F) -> DaftResult<common_treenode::VisitRecursion>
+    where
+        F: FnMut(&Self) -> DaftResult<common_treenode::VisitRecursion>,
+    {
+        for child in self.children().into_iter() {
+            match op(child.as_ref())? {
+                VisitRecursion::Continue => {}
+                VisitRecursion::Skip => return Ok(VisitRecursion::Continue),
+                VisitRecursion::Stop => return Ok(VisitRecursion::Stop),
+            }
+        }
+        Ok(VisitRecursion::Continue)
+    }
+
+    fn map_children<F>(self, transform: F) -> DaftResult<Self>
+    where
+        F: FnMut(Self) -> DaftResult<Self>,
+    {
+        let mut transform = transform;
+        let old_children = self.children();
+        let new_children = old_children
+            .iter()
+            .map(|c| transform(c.as_ref().clone()).map(Arc::new))
+            .collect::<DaftResult<Vec<_>>>()?;
+
+        if old_children
+            .into_iter()
+            .zip(new_children.iter())
+            .any(|(c1, c2)| c1 != c2)
+        {
+            // At least one child was transformed, so we need to create a new node.
+            Ok(self.with_new_children(new_children.as_slice()))
```

Comment:
I wonder if `transform` should be: `F: FnMut(&Self) -> DaftResult<Cow<Self>>`?

1. We can elide a clone `c.as_ref().clone()` when calling `transform`
2. `transform` can naively return the input ("no-op") or construct some new input and return as a `Cow`
3. In this function, we can make use of information from the `Cow` as a way of understanding if any of the transformations did anything. (`any(c is Owned for c in new_children)`) and then we need to call `self.with_new_children`

Reply:
Agreed, I started looking into doing this, but we'd need to align it with our current `Expr` representation, which already heavily uses the `TreeVisitor` trait to implement expression tree traversals and transformations, and currently takes ownership (move semantics). I decided to punt that for now since copying expression trees and logical/physical plans is not yet a bottleneck, and we need a more holistic solution there.

Reply:
Makes sense -- maybe we can add a comment in code that this is recognized as a potential performance issue.

This comment is non-blocking then